In [1]:
import pandas as pd
import datetime
import matplotlib.pyplot as plt
%matplotlib inline
plt.style.use('ggplot')
import seaborn as sns

sns.set_palette("deep", desat=0.6)
sns.set_context(rc={"figure.figsize": (8,4)})

In [14]:
dfAccidents = pd.read_csv('~/DATA/NYPD_Motor_Vehicle_Collisions.csv')

C:\Users\test\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py:2698: DtypeWarning: Columns (3) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [9]:
dfAccidents.head(8)

DATE  TIME    BOROUGH ZIP CODE   LATITUDE  LONGITUDE  \
0  03/13/2018  0:00   BROOKLYN    11207  40.678455 -73.896830   
1  03/13/2018  0:00   BROOKLYN    11216  40.678543 -73.949684   
2  03/13/2018  0:00   BROOKLYN    11230  40.619984 -73.964270   
3  03/13/2018  0:00   BROOKLYN    11235  40.579605 -73.964160   
4  03/13/2018  0:00  MANHATTAN    10002  40.719180 -73.990290   
5  03/13/2018  0:00  MANHATTAN    10012  40.726357 -73.993570   
6  03/13/2018  0:00  MANHATTAN    10022  40.755505 -73.968070   
7  03/13/2018  0:00        NaN      NaN  40.754677 -73.975815   

                  LOCATION                    ON STREET NAME  \
0   (40.678455, -73.89683)                               NaN   
1  (40.678543, -73.949684)  NOSTRAND AVENUE                    
2   (40.619984, -73.96427)  AVENUE L                           
3   (40.579605, -73.96416)                               NaN   
4    (40.71918, -73.99029)  ALLEN STREET                       
5   (40.726357, -73.99357)                               NaN   
6   (40.755505, -73.96807)  2 AVENUE                           
7  (40.754677, -73.975815)  EAST 46 STREET                     

     CROSS STREET NAME                           OFF STREET NAME  \
0                  NaN  90        JAMAICA AVENUE                   
1      ATLANTIC AVENUE                                       NaN   
2  CONEY ISLAND AVENUE                                       NaN   
3                  NaN  2955      BRIGHTON 4 STREET                
4      DELANCEY STREET                                       NaN   
5                  NaN  25        BOND STREET                      
6       EAST 51 STREET                                       NaN   
7                  NaN                                       NaN   

          ...           CONTRIBUTING FACTOR VEHICLE 2  \
0         ...                                     NaN   
1         ...                                     NaN   
2         ...                             Unspecified   
3         ...                             Unspecified   
4         ...                             Unspecified   
5         ...                             Unspecified   
6         ...          Driver Inattention/Distraction   
7         ...                         Other Vehicular   

   CONTRIBUTING FACTOR VEHICLE 3  CONTRIBUTING FACTOR VEHICLE 4  \
0                            NaN                            NaN   
1                            NaN                            NaN   
2                            NaN                            NaN   
3                            NaN                            NaN   
4                    Unspecified                            NaN   
5                            NaN                            NaN   
6                            NaN                            NaN   
7                            NaN                            NaN   

   CONTRIBUTING FACTOR VEHICLE 5  UNIQUE KEY            VEHICLE TYPE CODE 1  \
0                            NaN     3862869              PASSENGER VEHICLE   
1                            NaN     3862342  SPORT UTILITY / STATION WAGON   
2                            NaN     3863586  SPORT UTILITY / STATION WAGON   
3                            NaN     3862648              PASSENGER VEHICLE   
4                            NaN     3862181                           TAXI   
5                            NaN     3861800                             TK   
6                            NaN     3862159  SPORT UTILITY / STATION WAGON   
7                            NaN     3861487                           TAXI   

             VEHICLE TYPE CODE 2  VEHICLE TYPE CODE 3 VEHICLE TYPE CODE 4  \
0                            NaN                  NaN                 NaN   
1                            NaN                  NaN                 NaN   
2              PASSENGER VEHICLE                  NaN                 NaN   
3  SPORT UTILITY / STATION WAGON                  NaN                 NaN   
4  SPORT UTILITY /

In [10]:
dfAccidents.tail(8)

DATE  TIME    BOROUGH ZIP CODE   LATITUDE  LONGITUDE  \
1226366  07/01/2012  9:40     QUEENS    11354  40.764455 -73.831773   
1226367  07/01/2012  9:40     QUEENS    11369  40.767470 -73.881859   
1226368  07/01/2012  9:40     QUEENS    11370  40.758743 -73.890690   
1226369  07/01/2012  9:45        NaN      NaN        NaN        NaN   
1226370  07/01/2012  9:50  MANHATTAN    10019  40.766779 -73.996762   
1226371  07/01/2012  9:50     QUEENS    11001  40.736245 -73.702966   
1226372  07/01/2012  9:57  MANHATTAN    10065  40.765242 -73.957868   
1226373  07/01/2012  9:59      BRONX    10452  40.835397 -73.920305   

                          LOCATION                    ON STREET NAME  \
1226366   (40.7644548, -73.831773)  35 AVENUE                          
1226367  (40.7674704, -73.8818593)  23 AVENUE                          
1226368  (40.7587426, -73.8906895)  31 AVENUE                          
1226369                        NaN                               NaN   
1226370  (40.7667789, -73.9967618)  WEST SIDE HIGHWAY                  
1226371  (40.7362448, -73.7029656)  265 STREET                         
1226372  (40.7652424, -73.9578679)  1 AVENUE                           
1226373    (40.835397, -73.920305)  EAST 167 STREET                    

                        CROSS STREET NAME OFF STREET NAME         ...          \
1226366  FARRINGTON STREET                            NaN         ...           
1226367  88 STREET                                    NaN         ...           
1226368  77 STREET                                    NaN         ...           
1226369                               NaN             NaN         ...           
1226370  WEST 50 STREET                               NaN         ...           
1226371  85 AVENUE                                    NaN         ...           
1226372  EAST 68 STREET                               NaN         ...           
1226373  GERARD AVENUE                                NaN         ...           

         CONTRIBUTING FACTOR VEHICLE 2  CONTRIBUTING FACTOR VEHICLE 3  \
1226366                    Unspecified                            NaN   
1226367                    Unspecified                            NaN   
1226368                    Unspecified                            NaN   
1226369                    Unspecified                            NaN   
1226370                    Unspecified                            NaN   
1226371                    Unspecified                            NaN   
1226372                Other Vehicular                            NaN   
1226373                    Unspecified                            NaN   

         CONTRIBUTING FACTOR VEHICLE 4  CONTRIBUTING FACTOR VEHICLE 5  \
1226366                            NaN                            NaN   
1226367                            NaN                            NaN   
1226368                            NaN                            NaN   
1226369                            NaN                            NaN   
1226370                            NaN                            NaN   
1226371                            NaN                            NaN   
1226372                            NaN                            NaN   
1226373                            NaN                            NaN   

         UNIQUE KEY            VEHICLE TYPE CODE 1  \
1226366      245559              PASSENGER VEHICLE   
1226367      279641                  PICK-UP TRUCK   
1226368      279639  SPORT UTILITY / STATION WAGON   
1226369     2896726              PASSENGER VEHICLE   
1226370       37635                            VAN   
1226371      219395              PASSENGER VEHICLE   
1226372       44907              PASSENGER VEHICLE   
1226373       85154              PASSENGER VEHICLE   

                   VEHICLE TYPE CODE 2  VEHICLE TYPE CODE 3  \
1226366              PASSENGER VEHICLE                  NaN   
1226367  SPORT UTILITY / STATION WAGON                  NaN   
1226

In [15]:
dfAccidents.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1226374 entries, 0 to 1226373
Data columns (total 29 columns):
DATE                             1226374 non-null object
TIME                             1226374 non-null object
BOROUGH                          876463 non-null object
ZIP CODE                         876369 non-null object
LATITUDE                         1010457 non-null float64
LONGITUDE                        1010457 non-null float64
LOCATION                         1010457 non-null object
ON STREET NAME                   986184 non-null object
CROSS STREET NAME                918987 non-null object
OFF STREET NAME                  192420 non-null object
NUMBER OF PERSONS INJURED        1226374 non-null int64
NUMBER OF PERSONS KILLED         1226374 non-null int64
NUMBER OF PEDESTRIANS INJURED    1226374 non-null int64
NUMBER OF PEDESTRIANS KILLED     1226374 non-null int64
NUMBER OF CYCLIST INJURED        1226374 non-null int64
NUMBER OF CYCLIST KILLED         1226374

In [37]:

dfAccidents['DATENew'] =pd.to_datetime(dfAccidents['DATE'])


In [41]:
#dfAccidents['year'], dfAccidents['month'] = dfAccidents['DATENew'].dt.year, dfAccidents['DATENew'].dt.month
dfAccidents['year'], dfAccidents['month'],dfAccidents['DAY_OF_WEEK']= dfAccidents['DATENew'].apply(lambda x: x.year), dfAccidents['DATENew'].apply(lambda x: x.month), dfAccidents['DATENew'].apply(lambda x: x.weekday()) 


In [45]:
dfAccidents['year'].value_counts()


2017    229290
2016    227813
2015    217663
2014    206023
2013    203722
2012    100537
2018     41326
Name: year, dtype: int64

In [46]:
dfAccidents['month'].value_counts()

10    112336
7     110153
12    109372
9     108727
8     108585
11    107478
1     101817
5      96999
6      96133
3      95798
2      93298
4      85678
Name: month, dtype: int64

In [47]:
dfAccidents['DATE'].value_counts(dropna=False).head(20)

01/21/2014    1161
12/15/2017     994
05/19/2017     964
01/18/2015     961
02/03/2014     960
03/06/2015     939
05/18/2017     905
01/07/2017     895
12/14/2017     879
09/30/2016     872
03/02/2018     868
11/26/2013     867
04/28/2017     854
10/21/2016     851
03/08/2013     851
06/22/2017     840
10/27/2016     831
03/16/2017     831
03/05/2015     830
06/13/2017     822
Name: DATE, dtype: int64

Highest number of accidents happend on 01/21/2014    (1161) - interestingly it was Tuesday  

In [57]:
dfAccidents['DAY_OF_WEEK'].value_counts()

4    196304
3    183687
1    182072
2    180811
0    175771
5    163016
6    144713
Name: DAY_OF_WEEK, dtype: int64

## Time

In [60]:
dfAccidents['TIME'] = pd.to_datetime(dfAccidents['TIME'])                        

In [61]:
dfAccidents['hour'] =dfAccidents['TIME'].dt.hour    

In [62]:
dfAccidents['hour'].value_counts()

16    91503
17    88786
14    83935
18    78021
15    76289
13    72827
8     69244
12    68880
9     68332
11    65055
19    63547
10    63002
20    52628
21    43371
22    39097
7     34435
0     33098
23    31708
6     25223
1     19256
5     16079
4     14743
2     14717
3     12598
Name: hour, dtype: int64

Most of the accidents happen between 2pm and 6pm

In [65]:
dfAccidents['BOROUGH'].value_counts(dropna=True).head(20)

BROOKLYN         269626
QUEENS           229421
MANHATTAN        219974
BRONX            117255
STATEN ISLAND     40187
Name: BOROUGH, dtype: int64

In [66]:
dfAccidents.groupby(['BOROUGH', 'year']).size()

BOROUGH        year
BRONX          2012     9510
               2013    19768
               2014    19687
               2015    21252
               2016    21703
               2017    21187
               2018     4148
BROOKLYN       2012    23302
               2013    47017
               2014    47758
               2015    50843
               2016    47698
               2017    44633
               2018     8375
MANHATTAN      2012    20719
               2013    41578
               2014    41262
               2015    42518
               2016    36766
               2017    31488
               2018     5643
QUEENS         2012    19139
               2013    39310
               2014    41019
               2015    42484
               2016    41556
               2017    38378
               2018     7535
STATEN ISLAND  2012     4902
               2013     8295
               2014     6614
               2015     6344
               2016     6720
               2017    